File: leetcode/java/0003. Longest Substring Without Repeating Characters.java
package leetcode.java;

// Time complexity: O(N) N refers to the length of the input string
// Time complexity: O(M) M refers to the length of the longest substring

import java.util.HashSet;

class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashSet<Character> map = new HashSet<>();

        int left = 0;
        int right = 0;
        int maxLength = 0;

        while (right < s.length()) {
            if (map.contains(s.charAt(right))) {
                while (map.contains(s.charAt(right))) {
                    map.remove(s.charAt(left));
                    left += 1;
                }
            }
            map.add(s.charAt(right));
            maxLength = Integer.max(right - left + 1, maxLength);
            right += 1;
        }
        return maxLength;
    }
}

File: leetcode/python/0003. Longest Substring Without Repeating Characters.py
# Time complexity: O(N) N refers to the length of the input string
# Time complexity: O(M) M refers to the length of the longest substring

class Solution(object):
    def lengthOfLongestSubstring(self, s):
        if len(s) == 0:
            return 0
        elif len(s) == 1:
            return 1
        
        maxLength = 0
        left = 0
        right = 0
        
        storage = set()
        storage.add(s[left])

        while right < len(s)-1:
            right += 1
            while s[right] in storage:
                storage.remove(s[left])
                left += 1
            storage.add(s[right])
            maxLength = max(maxLength, right-left+1)
        return maxLength